File: 2015/day_2/main.py
import time

def main():
    with open("input.txt", "r") as f:
        input = f.read()

    start = time.time()

    input = input.split("\n")

    total_sqft_of_wrapping_paper = 0
    total_feet_of_ribbon = 0
    
    for line in input:
        if not line:
            continue

        # split input into the measurements and SORT ascendingly
        int_measurement = sorted([int(x) for x in line.split('x')])
        # calculation according to formulae
        surface_areas = [2 * int_measurement[0] * int_measurement[1], 2 * int_measurement[1] * int_measurement[2], 2 * int_measurement[0] * int_measurement[2]]

        sqft_of_wrapping_paper = sum(surface_areas) + int(min(surface_areas) / 2)
        total_sqft_of_wrapping_paper += sqft_of_wrapping_paper

        feet_of_ribbon = 2 * int_measurement[0] + 2 * int_measurement[1] + (int_measurement[0] * int_measurement[1] * int_measurement[2])
        total_feet_of_ribbon += feet_of_ribbon

    end = time.time()

    return total_sqft_of_wrapping_paper, total_feet_of_ribbon, end - start


if __name__ == "__main__":

    total_sqft_of_wrapping_paper, total_feet_of_ribbon, _time = main()

    print(f"The elves need {total_sqft_of_wrapping_paper} square feet of wrapping paper")  
    print(f"The elves need {total_feet_of_ribbon} feet of ribbon")  
    print(f"Needed {_time} seconds")

File: 2015/day_2/main.c
#include <stdio.h>
#include <string.h>
#include <time.h>
#include "line.h"

int main() {
    // open file, and verify it exists
    FILE *fp = fopen("input.txt", "r");

    if (!fp) {
        printf("No file found");
        return 1;
    }
    
    clock_t start = clock();

    int totalSqftOfWrappingPaper = 0;
    int totalFeetOfRibbon = 0;

    char *line = NULL;
    size_t len = 0;
    ssize_t read;

    char buffer[5];
    int surfaceAreas[] = {0, 0, 0};
    // w x l x h
    int values[] = {0, 0, 0};

    // read input line by line
    while ((read = getLine(&line, &len, fp)) != -1 ) {
        values[0] = 0;
        values[1] = 0;
        values[2] = 0;
        int previousX = 0;
        int lineLength = (int) strlen(line);
        // iterate over each line
        for (int i = 0; i < lineLength; i++) {
            // split line into single parts
            if (line[i] == 'x') {
                strncpy(buffer, line + previousX, i);
                previousX = i+1;
                if (!values[0]) {
                    values[0] = atoi(buffer);
                }
                else {
                    values[1] = atoi(buffer);
                }
            }

            if (values[0] && values[1]) {
                strncpy(buffer, line + previousX, lineLength);
                values[2] = atoi(buffer);
                break;
            }
        }

        // insertion sort algorithm, cause why not
        for (unsigned int i =1; i < (sizeof(values)/ sizeof(values[0])); i++ ) {
            int tmp = values[i];
            int j = i -1;
            while (j >= 0 && values[j] > tmp) {
                values[j+1] = values[j];
                j = j-1;
            }
            values[j+1] = tmp;
        }

        // formulas come from the description
        surfaceAreas[0] = 2 * values[0] * values[1];
        surfaceAreas[1] = 2 * values[0] * values[2];
        surfaceAreas[2] = 2 * values[1] * values[2];

        int minValue = -1;
        int sqftOfWrappingPaper = 0;
        int feetOfRibbon = 0;

        // find the minimal value as "slack"
        for (unsigned int i =0; i < 3; i++ ) {
            if (surfaceAreas[i] < minValue || minValue < 0) {
                minValue = surfaceAreas[i];
            }
            sqftOfWrappingPaper += surfaceAreas[i];
        }
        minValue = minValue / 2;

        sqftOfWrappingPaper += minValue;

        feetOfRibbon = 2 * values[0] + 2 * values[1] + (values[0] * values[1] * values[2]);

        totalSqftOfWrappingPaper += sqftOfWrappingPaper;
        totalFeetOfRibbon += feetOfRibbon;

        // Hallelujah for emptying buffers...
        memset(buffer,0,strlen(buffer));
    }

    clock_t end = clock();

    fclose(fp);
    free(line);

    printf("The elves need %d square feet of wrapping paper\n", totalSqftOfWrappingPaper);
    printf("The elves need %d feet of ribbon\n", totalFeetOfRibbon);
    float seconds = (float)(end - start) / CLOCKS_PER_SEC;
    printf("\nNeeded Time: %f seconds", seconds);

    return 0;
}